ab3d2_source/hireswall.s

```diff
@@ -1366,6 +1366,58 @@ nocliptop:
 
 ;wlcnt:			dc.w	0 ; unused
 
+				ifd OPT060
+
+drawwallS060_loop	macro
+val set \2
+.loop\<val>:
+				; grab packed texel
+				ifeq    \1-0
+				moveq   #%00011111,d1
+				and.b   1(a5,d4.w*2),d1
+				endc
+				ifeq    \1-1
+				move.w  (a5,d4.w*2),d1
+				lsr.w   #5,d1
+				and.w   #%00011111,d1
+				endc
+				ifeq    \1-2
+				moveq   #%01111100,d1
+				and.b   (a5,d4.w*2),d1
+				lsr.b   #2,d1
```

Comment:
Nice.

Reply:
is this approach still faster than bfextu?

Reply:
On 060 BFEXTU takes 6+ cycles (Table 10-16 of MC68060UM) and is pOEP-only (Table 10-2, meaning no other instruction can execute in the sOEP) so it's nearly always better to do use anything else (the other bitfield instructions are even worse).

Assuming the memory read is from cache case 0 can execute in 1 cycle, case 1 in 2.5, and case 2 in 1.5. In some of the cases the "half-cycles" will be wasted, and maybe there are few more cycles to be gained (e.g. `move.l #31<<5,d1 \ move.w ... \ lsr.w #5,d1` is probably slightly faster for case 1)

Reply:
Bitfield instructions are pure poison. They are perhaps faster than the set of instructions you would need on 020/030 to replicate their functionality but for something simple like this I think they'd be worse.

I think the correct answer to all this is to (eventually) do away with the packed texel format. Being able to use 256-colour textures would be a significant bonus. It's worth noting that both the floors and polygon object textures use typical 8-bit texels.

Reply:
Doing away with the packed texture format seems like a good idea. If I'm reading the 020/030 manuals correctly it won't increase performance (the opposite in fact), and on 060 (and probably 040) I don't think the potential cache benefits are worth it. Memory requirements will increase of course, but in 2023 I doubt that's an issue (even for "low-end" systems).

Reply:
I am a bit confused by your statement - you seem to be suggesting there won't be any benefits, only drawbacks. Not having to mask out the pixel from the texture data on the 020/030 when you can just access the appropriate byte directly, is a win, surely?

Reply:
@mheyer32 The original code also had some instruction cache freezing around some of these inner loops. Did we ever validate if they helped on slower CPUs?

Reply:
Yes, bad typo, meant "won't *decrease performance*". Not having to decode the packed format would INCREASE PERFORMANCE (I conjecture) even on 020/030 (at the cost of extra memory used).

Reply:
From what I understood: the intention of the instruction cache freeze at the very end of the first iteration is to keep the inner loop in cache for the remainder of the strip rendering. This way the only memory accesses should be the ones to fetch texture data and write out pixels. Maybe that helped on seriously constrained machines like an unexpanded A1200 or the CD32... but I have my doubts about fastram equipped machines(?)
Its definitely not helpful for machines with larger caches (040, 060) as we should naturally have enough cache to keep all involved instructions in there without manual "guidance".